pkg/operator/csi/controller/driver/controller.go

```diff
@@ -0,0 +1,374 @@
+// Liveness probe: csi-liveness-probe
+// Node driver registrar: csi-node-driver-registrar
+//
+// On every sync, this controller reads both the Deployment and the DaemonSet from static files and overrides a few fields:
+//
+// 1. Container image location
+//
+// The controller will replace the image specified in the static files if its respective environemnt variable is set. This
+// is a list of environment variables that the controller understands:
+//
+// DRIVER_IMAGE
+// PROVISIONER_IMAGE
+// ATTACHER_IMAGE
+// RESIZER_IMAGE
+// SNAPSHOTTER_IMAGE
+// NODE_DRIVER_REGISTRAR_IMAGE
+// LIVENESS_PROBE_IMAGE
+//
+// 2. Log level
+//
+// The controller can also override the log level passed in to the CSI driver container.
+//
+// In order to do that, the flag `--v=` passed in to the csi-driver container is replaced with value specified
+// in the OperatorClient resource (Spec.LogLevel). That being said, it's expected that the csi-driver container sets
+// its log level based on the `--v=` flag, which is initialized by default if the driver uses the package klog.
+type Controller struct {
+	// Controller
+	name            string
+	queue           workqueue.RateLimitingInterface
+	eventRecorder   events.Recorder
+	informersSynced []cache.InformerSynced
+
+	// CSI driver
+	csiDriverName      string
+	csiDriverNamespace string
+	assetFunc          func(string) []byte
+	controllerManifest []byte
+	nodeManifest       []byte
+	images             images
+
+	// Resources
+	kubeClient     kubernetes.Interface
+	operatorClient v1helpers.OperatorClient
+}
+
+type images struct {
+	csiDriver           string
+	attacher            string
+	provisioner         string
+	resizer             string
+	snapshotter         string
+	nodeDriverRegistrar string
+	livenessProbe       string
+}
+
+// New returns a new Controller without any CSI Service.
+// It's expected that the CSI Services are added to the controller returned by
+// this function using Service's respective factory function.
+func New(
+	name string,
+	csiDriverName string,
+	csiDriverNamespace string,
+	operatorClient v1helpers.OperatorClient,
+	assetFunc func(string) []byte,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) *Controller {
+	controller := &Controller{
+		name:               name,
+		csiDriverName:      csiDriverName,
+		csiDriverNamespace: csiDriverNamespace,
+		kubeClient:         kubeClient,
+		operatorClient:     operatorClient,
+		assetFunc:          assetFunc,
+		images:             imagesFromEnv(),
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), csiDriverName),
+		eventRecorder:      eventRecorder,
+	}
+	operatorClient.Informer().AddEventHandler(controller.eventHandler(csiDriverName))
+	controller.informersSynced = append(controller.informersSynced, operatorClient.Informer().HasSynced)
+	return controller
+}
+
+// WithControllerService returns a Controller with the CSI Controller Service.
```

Comment:
godoc needs greater specificity about this being required to be a deployment.

Reply:
Added